File: scripts/Remove Duplicate Letters/Remove Duplicate Letters.py
class Solution:
    def removeDuplicateLetters(self, s: str) -> str:
        
		last_occ = {}
		stack = []
		visited = set()

		for i in range(len(s)):
			last_occ[s[i]] = i

		for i in range(len(s)):

			if s[i] not in visited:
				while (stack and stack[-1] > s[i] and last_occ[stack[-1]] > i):
					visited.remove(stack.pop())

				stack.append(s[i])
				visited.add(s[i])

		return ''.join(stack)

File: scripts/algorithms/R/Remove Duplicate Letters/Remove Duplicate Letters.js
// Runtime: 56 ms (Top 86.81%) | Memory: 51.10 MB (Top 5.49%)

/**
 * @param {string} s
 * @return {string}
 */
var removeDuplicateLetters = function(s) {
    const lastOccurrence = {};
    for (let i = 0; i < s.length; i++) {
        lastOccurrence[s[i]] = i;
    }

    const stack = [];
    const visited = new Set();

    for (let i = 0; i < s.length; i++) {
        if (visited.has(s[i])) {
            continue;
        }

        while (
            stack.length > 0 &&
            s[i] < stack[stack.length - 1] &&
            i < lastOccurrence[stack[stack.length - 1]]
        ) {
            visited.delete(stack.pop());
        }

        visited.add(s[i]);
        stack.push(s[i]);
    }

    return stack.join('');    
};
